services/src/main/java/org/keycloak/services/util/DPoPUtil.java

```diff
@@ -0,0 +1,401 @@
+                new DPoPIsActiveCheck(session, lifetime, clockSkew),
+                new DPoPReplayCheck(session, lifetime));
+
+        if (accessToken != null) {
+            verifier.withChecks(new DPoPAccessTokenHashCheck(accessToken));
+        }
+
+        try {
+            DPoP dPoP = verifier.verify().getToken();
+            dPoP.setThumbprint(JWKSUtils.computeThumbprint(jwk));
+            return dPoP;
+        } catch (DPoPVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("DPoP verification failure", ex);
+        }
+
+    }
+
+    public static void validateBinding(AccessToken token, DPoP dPoP) throws VerificationException {
+        try {
+            TokenVerifier.createWithoutSignature(token)
+                    .withChecks(new DPoPUtil.DPoPBindingCheck(dPoP))
+                    .verify();
+        } catch (TokenVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("Token verification failure", ex);
+        }
+    }
+
+    public static void bindToken(AccessToken token, String thumbprint) {
+        AccessToken.Confirmation confirmation = token.getConfirmation();
+
+        if (confirmation == null) {
+            confirmation = new AccessToken.Confirmation();
+            token.setConfirmation(confirmation);
+        }
+
+        confirmation.setKeyThumbprint(thumbprint);
+    }
+
+    public static void bindToken(AccessToken token, DPoP dPoP) {
+        bindToken(token, dPoP.getThumbprint());
+    }
+
+    private static class DPoPClaimsCheck implements TokenVerifier.Predicate<DPoP> {
+
+        static final TokenVerifier.Predicate<DPoP> INSTANCE = new DPoPClaimsCheck();
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            Long iat = t.getIat();
+            String jti = t.getId();
+            String htu = t.getHttpUri();
+            String htm = t.getHttpMethod();
+
+            if (iat != null &&
+                isNotBlank(jti) &&
+                isNotBlank(htm) &&
+                isNotBlank(htu)) {
+                return true;
+            } else {
+                throw new DPoPVerificationException(t, "DPoP mandatory claims are missing");
+            }
+        }
+
+    }
+
+    private static class DPoPHTTPCheck implements TokenVerifier.Predicate<DPoP> {
+
+        private final URI uri;
+        private final String method;
+
+        DPoPHTTPCheck(URI uri, String method) {
+            this.uri = uri;
+            this.method = method;
+        }
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            try {
+                if (!normalize(new URI(t.getHttpUri())).equals(normalize(uri)))
+                    throw new DPoPVerificationException(t, "DPoP HTTP URL mismatch");
+
+                if (!method.equals(t.getHttpMethod()))
```

Comment:
AFAIK, http method is case sensitive. How should we do if a client sends a DPoP proof including "htm" values including lower case? (e.g, "htm": "post")